### DISCUSSION TITLE
Welcome to FastAPI Discussions!

### DISCUSSION THREAD

**Discussion by tiangolo** [fastapi/fastapi#2475] [Questions] [answered]
## 👋 Welcome!

We’re using Discussions as a place to connect with other members of our community. We hope that you:

* Help others with your FastAPI knowledge. 🤓
* Ask questions you’re wondering about. ❓
* Share ideas. 💡
* Share what you do with FastAPI. 🚀
* Engage with other community members. 👥
* Welcome others and are open-minded. ⭐

To get started, comment below with an introduction of yourself and tell us about what you do with FastAPI.

**Comment by Kludex** [comment]
Hi there! :wave:

I'm a :brazil: software engineer currently living in :fr:.

I :heart:  FastAPI and I've been using it in my professional life, and in the personal as well, working on some auxiliary packages and helping the community! :rocket:

**Reply by Kludex** [reply]
Hi there! 👋 

It's been awhile. I'm not in 🇫🇷  anymore, I'm in the 🇳🇱 .

I still ❤️ FastAPI, and I'm now a maintainer of Uvicorn and Starlette. 🙌

**Reply by Kludex** [reply]
A year later. I'm still around.

**Comment by cclinet** [comment]
I'm using fastAPI for serving my ML models in many projects this year
Hope I can make some contributions

**Comment by hard-coders** [comment]
🚀 It's great!
May I open a discussion for 🇰🇷 Korean translation?
I want to discuss with people who want to translate Korean and talk about writing style, glossaries, and any stuff freely.

**Comment by jypara** [comment]
Hey hey, I am very new with fastapi and I am having some issues with api and ajax.
I am getting error 405

Ajax function

```
function submitForm() {
        var url = 'http://localhost:8000';
        let id = document.getElementById("person-id").value;
        var data = {id: id};
        fetch(url, {
          method: 'PUT',
          body: JSON.stringify(data),
          headers:{
            'Content-Type': 'application/json'
          }
        }).then(res => res.json())
        .then(response => console.log('Success:', JSON.stringify(response)))
        .catch(error => console.log('Error:', error));
```
Here py method update
```
@router.put("/{id}",  response_class=Response)
async def update(request: Request, id: str, req: UpdateModel = Body(...)):
   updated = await update(id, req.dict())
   return templates.TemplateResponse('index.html', context={'request': request})
```
Update is not working, console is just showing

405 method not allowed T

**Reply by ycd** [reply]
Please open an issue.

**Comment by udit-pandey** [comment]
Hi. I have implemented a websocket using fastapi. I am load testing it using Jmeter websocket sampler and some of these requests are failing with WebSocket I/O error: Read timed out (at least this is the error that I get on Jmeter side). The error on server side:

```
import json

import uvicorn
from fastapi import FastAPI, Request
from fastapi.middleware.cors import CORSMiddleware
from starlette.websockets import WebSocket


app = FastAPI()



@app.get("/hello")
def index():
    return {"message": "hello"}


@app.websocket("/post")
async def post_req(websocket: WebSocket):
    await websocket.accept()
    while True:
        body = await websocket.receive_json()
        print(body)
        await websocket.send_text("some text")


if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8443)

```

```

ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/home/udit.pandey/Desktop/workspace/chat_server/venv/lib/python3.7/site-packages/uvicorn/protocols/websockets/websockets_impl.py", line 157, in run_asgi
    result = await self.app(self.scope, self.asgi_receive, self.asgi_send)
  File "/home/udit.pandey/Desktop/workspace/chat_server/venv/lib/python3.7/site-packages/uvicorn/middleware/proxy_headers.py", line 45, in __call__
    return await self.app(scope, receive, send)
  File "/home/udit.pandey/Desktop/workspace/chat_server/venv/lib/python3.7/site-packages/fastapi/applications.py", line 199, in __call__
    await super().__call__(scope, receive, send)
  File "/home/udit.pandey/Desktop/workspace/chat_server/venv/lib/python3.7/site-packages/starlette/applications.py", line 111, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/home/udit.pandey/Desktop/workspace/chat_server/venv/lib/python3.7/site-packages/starlette/middleware/errors.py", line 146, in __call__
    await self.app(scope, receive, send)
  File "/home/udit.pandey/Desktop/workspace/chat_server/venv/lib/python3.7/site-packages/starlette/exceptions.py", line 58, in __call__
    await self.app(scope, receive, send)
  File "/home/udit.pandey/Desktop/workspace/chat_server/venv/lib/python3.7/site-packages/starlette/routing.py", line 566, in __call__
    await route.handle(scope, receive, send)
  File "/home/udit.pandey/Desktop/workspace/chat_server/venv/lib/python3.7/site-packages/starlette/routing.py", line 283, in handle
    await self.app(scope, receive, send)
  File "/home/udit.pandey/Desktop/workspace/chat_server/venv/lib/python3.7/site-packages/starlette/routing.py", line 57, in app
    await func(session)
  File "/home/udit.pandey/Desktop/workspace/chat_server/venv/lib/python3.7/site-packages/fastapi/routing.py", line 247, in app
    await dependant.call(**values)
  File "./websocket.py", line 29, in post_req
    body = await websocket.receive_json()
  File "/home/udit.pandey/Desktop/workspace/chat_server/venv/lib/python3.7/site-packages/starlette/websockets.py", line 98, in receive_json
    self._raise_on_disconnect(message)
  File "/home/udit.pandey/Desktop/workspace/chat_server/venv/lib/python3.7/site-packages/starlette/websockets.py", line 80, in _raise_on_disconnect
    raise WebSocketDisconnect(message["code"])
starlette.websockets.WebSocketDisconnect: 1006

```

There are 2 scenarios that I have covered with this:
1. 5 users spawned within 10sec
    Result: No failures. 
2. 10 users spawned within 10sec
    Result: Random failures with error "WebSocket I/O error: Read timed out" on Jmeter sampler.

Any help on this is greatly appreciated. Thanks.

**Comment by turing80** [comment]


**Reply by Kludex** [reply]
This is not the right place :disappointed: 

Do you mind creating an issue (question)? https://github.com/tiangolo/fastapi/issues/new/choose

**Comment by walkershashi** [comment]
Hi, 
I am new to FastAPI, and I am not able to figure out how to pass the form data from the HTML page containing forms into the backend or to the other HTML page.

I have created an endpoint to serve the form which takes input, and I want those input to be displayed on the terminal or on the new HTML page served by another endpoint.
Can anyone help me out with this?

Thanks for your help.

**Reply by Mause** [reply]
Have you tried following the documentation for forms? See https://fastapi.tiangolo.com/tutorial/request-forms-and-files/

**Reply by walkershashi** [reply]
Yes, but it doesn't address the issue I am looking for.

I have a form and want its data to be transferred to another HTML page. How do I do that using fast API

**Reply by Mause** [reply]
If you have an issue, please create it on GitHub here:  https://github.com/tiangolo/fastapi/issues/new/choose

**Comment by Kamuzinzi** [comment]
Hello!
I am trying to create a Facebook chatbot using FastAPI but I cannot find how I can connect FastAPI to Facebook. 
Is there anybody who knows how to do it?

**Comment by giridhart** [comment]
Hi,
Trying to get something work quickly with FastAPI,  I have started with `full-stack-fastapi-postgresql` I thought it would be easy to modify to my needs but it is quite overwhelming.
First change I wan to make is change the authentication to Google Oauth or add Google Oauth as and extra for accessing some parts of the website. Where should I start? 
Thank you.

**Comment by claudioalberto07** [comment]
Hello, devs. 
I got the following error after deploying through Lambda Function and creating API Gateway on AWS. I haven't found anything like it yet in the community.
From what it looks like something like the AWS proxy blocking the Swagger route. Has anyone got this error?

![image](https://user-images.githubusercontent.com/72947654/159194800-00191d50-2d1b-4e23-862b-6e7f47fc15eb.png)

Thanks, guys.

**Comment by prashantkr006** [comment]
Hi tthere my query is quite simple but still confusing for me so the issue I am facing is that I am trying to retrive data but my data is coming from relationships tables also so in response body I am getting only the parent table data but not the data from relatinships table as well so how can I SOLVE this issue?
https://stackoverflow.com/questions/72840483/how-to-response-the-relationship-data-in-fastapi
here is the complete question in details

**Comment by nickolas1969** [comment]
New FastAPI user trying to read a post request that contains Basic Auth, a header "client_id", and a json body. 

Using bits of code from https://fastapi.tiangolo.com/tutorial I got BasicAuth to work, and I got Request to work to grab the headers and body. Headers also appears to get the username and password too, but I don't have access to them without the Authentication part.

```
@app.post("/home")
async def root(request: Request)
   my_id = request.headers.get('client_id')
   my_user = request.user

AssertionError: AuthenticationMiddleware must be installed to access request.user

```
If I just use root(credentials: HTTPBasicCredentials = Depends(security)) then I can get the user and password.

I was hoping to find an example how to get all 3 parts, or I am misunderstanding what it might be called to get them?

**Comment by alepadovam** [comment]
Hi there!
I have a web application in production to provide newborn screening results to authorized entities in the field of public health. 
I'm liking FastAPI and I want to inside this group because sometimes I  need helped. I'd like to help with what I've learned so far in this framework.
At the moment I've some problems to provide a response model from subclasses. Someone help me please?

**Reply by prashantkr006** [reply]
What are the problems you facing, may I know?

**Comment by Master-Servis** [comment]
Hi, maybe can someobe share some FasstAPI / vanilla typescript resources? Set up/git repos etc?

**Reply by prashantkr006** [reply]
The best resource is fastAPI documentation 
[fastAPI docs](https://fastapi.tiangolo.com/)

**Comment by tiangolo** [comment, marked as answer]
As this discussion was being used mainly to ask questions and not intros, I moved it to questions, to remove the "General" Discussions category and leave mainly the "Questions" category. :sweat_smile:

**Reply by HashimThePassionate** [reply]
Why developers switch to SQLModel instead of SQLAlchemy in 2024?

**Comment by veonim** [comment]
What is would be the best testing framework for FastAPI + SQLModel + Alembic application?

**Comment by Abhijith-Dev01** [comment]
How to apply foreign key in fastapi models.py file using SQlAlchemy

**Reply by rafveraldi** [reply]
You can refer to a foreign key by importing ForeignKey from SQLAlchemy in your models.py

See example below:

```
from sqlalchemy import Column, ForeignKey, Integer
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()

class Foo(Base):
    __tablename__ = "foos"

    id = Column(Integer, primary_key=True, index=True)

class Bar(Base):
    __tablename__ = "bars"

    id = Column(Integer, primary_key=True, index=True)
    foo_id = Column(Integer, ForeignKey("foos.id"))
```

**Comment by PranathiSG** [comment]
Hi, Does FastAPI support any data engineering libraries other than pandas that can be used efficiently and in an optimized way to process the data in FastAPI application ? FastAPI application will be deployed in kubernetes environment. 

We are dealing with huge number of data (let's say around 50,000 records per API request) that will be processed, transformed to the required schema format in our FastAPI application.

**Comment by flypenguin** [comment]
Hi all! For me it would be quite practical right now to process an _arbitrary_ form from a FastAPI method, like this:

```python
@app.post("/")
def process_any_form( ... ):
    pass
```

The problem is – how do I specify it (the `(...)` part of the code above ...)? When I look [into the docs](https://fastapi.tiangolo.com/tutorial/request-forms/), I need to specify all fields of the form. What do I do if I don't know them in advance?

Thanks for any hints!

**Comment by shruthim16** [comment]
Hi all, We have implemented fast api application with pydantic v1 and with release of pydantic v2 we were able to migrate using bymp-pydantic but we have used OneLogin package and imported OneLogin_Saml2_Auth , there are pydantic errors from this package , how do we resolve this ?

Any help , would be much appreciated Thank you!

**Reply by iudeen** [reply]
Ask Onelogin repo, may be?

**Comment by shashiwtt** [comment]
Hi all,
i'm facing CORS error  in my Application with FastAPI
But it's working in Postman and Web
i also added

like this 

app = FastAPI()

app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],  # Specify the allowed origin
    allow_credentials=True,
    allow_methods=["POST", "GET"],
    allow_headers=["*"],
    max_age=3600,
)

and added below way and many other ways but not working
app.add_middleware(CORSMiddleware,
                    allow_origins=["*"],  # Replace "*" with the list of domains you want to allow.
                    allow_credentials=True,
                    allow_methods=["*"],
                    allow_headers=["*"],
                )
Any help , would be much appreciated Thank you!

**Comment by xwzwx** [comment]
Actually I haven't much idea what is FastAPI and Why use it.
So I am gonna start with researching what is special in websites built using FastAPI.
If there is anyone who is gonna help poor newbie, I would appreciate it much!
Anyone can provide some links of websites built using FastAPI?

**Comment by prashantkr006** [comment]
@oscarcheerup Hey there, it is great to know that you're excited about learning FastAPI, if you are looking for any websites built using FastAPI then let me tell you that, FastAPI is a backend technology that is used to develop servers for websites, and to learn about it I will recommend the official documentation, they have structured it in pretty good format such that you will find everything easy and learning will be smooth, here you go [fastapi docs](https://fastapi.tiangolo.com/).

**Comment by sonustar** [comment]
hello guys , actually I am working in a project and I couldn't figure out . How to do that stuff .  frontend is next js and backend is fast API so my question is I am storing some responses in this format 
responses = {
    "What is apple ?": " Apple is a red coloured fruit ."
}

How can I fetch it in a rich text editor format in frontend ? What libraries should I use ??

**Comment by SanjinDedic** [comment]
# 👋 Hello, FastAPI Community!

I'm Sanjin, and I've been using FastAPI for two years to build backends for a statewide coding competition in Melbourne, Australia. 🇦🇺 So far, over 5,000 students have used it, and the backend has held up great! 💪

## 🚀 This Year's Ambitious Setup

We're running an exciting project with these features:

- 🤖 Students submit code agents that compete in games like Prisoner's Dilemma
- 🐳 Code runs safely in Docker containers
- 🗃️ Database is in SQLModel and running smoothly
- 🔌 Modular game engine design for super easy addition of new game types 
- ⚛️ Front-end is React (not pretty, but functional)

## 🔗 Check It Out!

You can find the project here: [Agent Games on GitHub](https://github.com/SanjinDedic/agent_games)

## 🙏 Feedback and Collaboration Welcome

I'd love any feedback on the project! The full-stack app takes just 10 minutes to set up locally. There are usually a dozen issues you can take on if you're interested in collaborating as well as an opportunity to create much cooler games than we came up with so far!

Let's build something amazing together! 🌟
